### PR TITLE
Add rbac support for spec alignment

### PR DESCRIPTION
**Pull request by bamachrn**
### Motivation
https://github.com/redhat-developer/service-binding-operator/issues/900

### Changes
* Added Aggregated cluster role for the operator to get access to the resources
* Added rolebindings for the openshift-operators and operators namespace

### Testing

For testing this we need to 
* re-create the bundle manifests
* Deploy the operator and bundle to the cluster

For further more details refer the CONTRIBUTING.md

### CODE REVIEW

File: controllers/servicebinding_controller.go [hunk 1]
```diff
@@ -49,7 +49,6 @@ type ServiceBindingReconciler struct {
 // +kubebuilder:rbac:groups="",resources=pods;services;endpoints;persistentvolumeclaims;events;configmaps;secrets,verbs=*
 // +kubebuilder:rbac:groups=apps,resources=deployments;daemonsets;replicasets;statefulsets,verbs=*
 // +kubebuilder:rbac:groups=apiextensions.k8s.io,resources=customresourcedefinitions,verbs=get;list;watch
-// +kubebuilder:rbac:groups=*,resources=*,verbs=get;list;watch;update
```
Comment (pedjak) [inline controllers/servicebinding_controller.go:52]: do we really need to touch this here? IMHO, that should addressed in https://github.com/redhat-developer/service-binding-operator/issues/454

**Review by pedjak — CHANGES_REQUESTED** [review]
we need acceptance test(s) checking that the added role is available in the cluster and that the rolebinding is established to the service account of the operator deployment.

File: controllers/servicebinding_controller.go (hunk 1 above)
Comment (bamachrn) [reply controllers/servicebinding_controller.go:52]: yes. make sense. I was thinking both the roles and access should go together.

File: test/acceptance/features/steps/steps.py [hunk 2]
```diff
@@ -576,3 +576,37 @@ def update_binding_secret(context):
...
     result = re.search(rf'.*{context.sb_secret}.*(created|unchanged|configured)', output)
     assert result is not None, f"Unable to apply YAML for binding secret '{context.sb_secret}': {output}"
+
+
+@then(u'cluster role "{cluster_role_name}" is available in the cluster')
+def check_cluster_role_exists(context, cluster_role_name):
+    openshift = Openshift()
+    roles_present, exit_code = openshift.is_resource_in("clusterroles")
+    if exit_code==0:
+        role_list = json.load(roles_present)
```
Comment (pmacik) [inline test/acceptance/features/steps/steps.py:586]: `roles_present` is actually a string containig output of `kubectl get clusterroles` - see the `openshift.is_role_in` function.
Furthermore, `json.load` expects a file not a string and so this line causes the acceptance tests to [fail](https://github.com/redhat-developer/service-binding-operator/pull/979/checks?check_run_id=3042889774#step:8:123) with the following:
```
Scenario: Service Binding Operator installation should contain aggregating cluster role and are bound with Namespace.  # test/acceptance/features/addRBACForOperator.feature:2
    Given Namespace [TEST_NAMESPACE] is used                                                                             # test/acceptance/features/steps/steps.py:43
    When Service Binding Operator is running                                                                             # test/acceptance/features/steps/steps.py:83
    Then cluster role "servicebinding-controller-role" is available in the cluster                                       # test/acceptance/features/steps/steps.py:590
      Traceback (most recent call last):
        File "/home/runner/work/service-binding-operator/service-binding-operator/out/venv3/lib/python3.7/site-packages/behave/model.py", line 1329, in run
          match.run(runner.context)
        File "/home/runner/work/service-binding-operator/service-binding-operator/out/venv3/lib/python3.7/site-packages/behave/matchers.py", line 98, in run
          self.func(context, *args, **kwargs)
        File "test/acceptance/features/steps/steps.py", line 595, in check_cluster_role_exists
          role_list = json.load(roles_present)
        File "/opt/hostedtoolcache/Python/3.7.10/x64/lib/python3.7/json/__init__.py", line 293, in load
          return loads(fp.read(),
      AttributeError: 'str' object has no attribute 'read'
      
      Captured stdout:
      TEST_NAMESPACE = test-namespace-772a9d798240
      ,---------,-
      | COMMAND : kubectl get ns test-namespace-772a9d798240
      '---------'-
      Namespace test-namespace-772a9d798240 is created!!!
      ,---------,-
      | COMMAND : kubectl get servicebinding
      '---------'-
      ,---------,-
      | COMMAND : kubectl get deployments --all-namespaces -o json | jq -rc '.items[] | select(.metadata.name == "service-binding-operator").metadata.namespace'
      '---------'-
      Service binding operator is running!!!
      ,---------,-
      | COMMAND : kubectl get clusterroles
      '---------'-
```

File: test/acceptance/features/steps/steps.py [hunk 3]
```diff
@@ -576,3 +576,37 @@ def update_binding_secret(context):
...
+def check_cluster_role_exists(context, cluster_role_name):
+    openshift = Openshift()
+    roles_present, exit_code = openshift.is_resource_in("clusterroles")
+    if exit_code==0:
+        if cluster_role_name in roles_present:
+            assert True, "Found the cluster role in the cluster"
+        else:
+            assert False, "Could not find the cluster roler '{cluster_role_name}'"
+    else:
+        assert False, "Cluster role does not exist in the cluster"
```
[the remark this answers is not in the document]
Comment (pmacik) [reply test/acceptance/features/steps/steps.py:591]: Please use asserts properly (https://www.w3schools.com/python/ref_keyword_assert.asp)

File: test/acceptance/features/steps/steps.py [hunk 4]
```diff
@@ -576,3 +576,37 @@ def update_binding_secret(context):
...
+           subjects_list = rolebinding_json.subjects
+           for i in len(subjects_list):
+                if subjects_list[i].kind == "ServiceAccount" and subjects_list[i].name == "servicebinding-operator":
+                   assert True
+                else:
+                    continue
+        else:
+            assert False, "Could not find rolebing for the role '{cluster_role_name}'"
+    else:
+        assert False, "Could not find the rolebinding in the cluster"
```
[the remark this answers is not in the document]
Comment (pmacik) [reply test/acceptance/features/steps/steps.py:612]: Please use asserts properly (https://www.w3schools.com/python/ref_keyword_assert.asp)

File: test/acceptance/features/addRBACForOperator.feature [hunk 5]
```diff
@@ -0,0 +1,7 @@
+Feature: Making Service Biniding Operator RBAC pluggable so that other controllers/admins can add additional rules for the service binding operator.
+    @rbactest
```
Comment (pedjak) [inline test/acceptance/features/addRBACForOperator.feature:2]: do we need an extra annotation here?

File: test/acceptance/features/steps/openshift.py [hunk 6]
```diff
@@ -146,7 +146,7 @@ def search_resource_in_namespace(self, resource_plural, name_pattern, namespace)
 
     def is_resource_in(self, resource_type):
         output, exit_code = self.cmd.run(f'{ctx.cli} get {resource_type}')
-        return exit_code == 0
+        return (output, exit_code)
```
Comment (pedjak) [inline test/acceptance/features/steps/openshift.py:149]: we changed here fully return values, so `is_resource_in` is not a proper name anymore.. btw, why do we need to change the return value?

File: test/acceptance/features/steps/openshift.py [hunk 7]
```diff
@@ -300,6 +300,10 @@ def get_resource_info_by_jsonpath(self, resource_type, name, namespace, json_pat
             print(f'Error getting value for {resource_type}/{name} in {namespace} path={json_path}: {output}')
             return None
 
+    def get_json_resource(self, resource_type, name, namespace):
+        output, exit_code = self.cmd.run(f'{ctx.cli} get {resource_type} {name} -n {namespace} -o json')
+        return output, exit_code
```
Comment (pedjak) [inline test/acceptance/features/steps/openshift.py:305]: why we are returning exit code? also, what we return here is just raw representation, perhaps we could already parse it and return json data structure?

I would definitely assert exist for 0, not return it to the caller.

File: test/acceptance/features/steps/servicebindingoperator.py [hunk 8]
```diff
@@ -16,7 +16,9 @@ def __init__(self,  name="service-binding-operator", namespace="openshift-operat
 
     def check_crd(self):
         crd_name = "servicebinding"
-        assert self.openshift.is_resource_in(crd_name), f"CRD '{crd_name}' does not exist"
```
Comment (pedjak) [inline test/acceptance/features/steps/servicebindingoperator.py:19]: the removed line what doing exactly what the added lines are doing now. What is the motivation for the change?

File: test/acceptance/features/steps/steps.py [hunk 9]
```diff
@@ -436,3 +437,34 @@ def update_binding_secret(context):
     output = openshift.apply(yaml.dump(secret_yaml), context.namespace.name)
     result = re.search(rf'.*{context.sb_secret}.*(created|unchanged|configured)', output)
     assert result is not None, f"Unable to apply YAML for binding secret '{context.sb_secret}': {output}"
+
+
+@then(u'cluster role "{cluster_role_name}" is available in the cluster')
+def check_cluster_role_exists(context, cluster_role_name):
+    openshift = Openshift()
+    roles_present, exit_code = openshift.is_resource_in("clusterroles")
```
Comment (pedjak) [inline test/acceptance/features/steps/steps.py:445]: instead of getting all cluster roles, and then checking if `cluster_role_name` is present, we can try to get read directly `cluster_role_name`. If not found, we are going to get exit code different from 0.

you can use here `openshift.get_json_resource` instead.

File: test/acceptance/features/steps/steps.py [hunk 10]
```diff
@@ -436,3 +437,34 @@ def update_binding_secret(context):
...
+    else:
+        assert False, "Cluster role does not exist in the cluster"
+
+
+@then(u'operator service account is bound to "{cluster_role_name}" cluster role')
+def check_service_account_bound(context, cluster_role_name):
+    openshift = Openshift()
+    sbr = Servicebindingoperator()
+    operator_namespace = sbr.check_deployment()
+    (rolebinding, exitcode) = openshift.get_json_resource("clusterrolebinding", "service-binding-operator-controller-rolebinding", operator_namespace)
```
Comment (pedjak) [inline test/acceptance/features/steps/steps.py:457]: see my earlier comment - the code can be simplified if assert exit code in `get_json_resource`

File: test/acceptance/features/steps/steps.py [hunk 11]
```diff
@@ -436,3 +437,34 @@ def update_binding_secret(context):
...
+    if exitcode == 0:
+        rolebinding_json = json.loads(rolebinding)
+        if rolebinding_json["roleRef"]["name"] == cluster_role_name:
+            subjects_list = rolebinding_json["subjects"]
+            for subject in subjects_list:
+                if subject["kind"] == "ServiceAccount" and subject["name"] == "service-binding-operator":
```
Comment (pedjak) [inline test/acceptance/features/steps/steps.py:463]: we should not hard code service account to `service-binding-operator`, instead we should ready from SBO deployment what is the service account in use and check against it.

**Review by pedjak — CHANGES_REQUESTED** [review]

File: test/acceptance/features/addRBACForOperator.feature (hunk 5 above)
Comment (bamachrn) [reply test/acceptance/features/addRBACForOperator.feature:2]: no this was for running the tests locally. reverting

File: test/acceptance/features/steps/openshift.py (hunk 6 above)
Comment (bamachrn) [reply test/acceptance/features/steps/openshift.py:149]: this function checks if the resource type exists in the cluster. Here we want to make sure if a specific resource exists in the cluster irrespective of the namespace. I suppose we can use resource name directly here without changing the return value as well.

File: test/acceptance/features/steps/openshift.py (hunk 7 above)
Comment (bamachrn) [reply test/acceptance/features/steps/openshift.py:305]: Here we are trying to handle use cases where there are multiple entries for a specific resource type.
Like when we are checking for the `rolebindings` we can have multiple rules/ subjects for a single rolebinding. With the exit code we can make sure if the resource is present in the cluster. and with the json output we can check for different values in the step implementation.

File: test/acceptance/features/steps/servicebindingoperator.py (hunk 8 above)
Comment (bamachrn) [reply test/acceptance/features/steps/servicebindingoperator.py:19]: we changed the return type for the `is_resource_in` function. that's why updated this function to assert on the proper value.

File: test/acceptance/features/steps/steps.py (hunk 11 above)
Comment (bamachrn) [reply test/acceptance/features/steps/steps.py:463]: ack

File: config/rbac/servicebinding_controller_rolebinding.yaml [hunk 12]
```diff
@@ -0,0 +1,15 @@
+apiVersion: rbac.authorization.k8s.io/v1
+kind: ClusterRoleBinding
+metadata:
+  name: controller-rolebinding
+roleRef:
+  apiGroup: rbac.authorization.k8s.io
+  kind: ClusterRole
+  name: service-binding-operator-controller-role
```
Comment (baijum) [inline config/rbac/servicebinding_controller_rolebinding.yaml:8]: Where is this ClusterRole defined?

Comment (bamachrn) [reply config/rbac/servicebinding_controller_rolebinding.yaml:8]: we are adding this ClusterRole as an aggregator ClusterRole, to make it pluggable for other controllers to add additional rules. We are adding this to manifests here: https://github.com/redhat-developer/service-binding-operator/pull/979/files#diff-3fd064f2f1027a954b035a6d5885b96cc5b483f04889d14fa840ace5e9c29c55R8-R9

File: config/rbac/servicebinding_controller_role.yaml [hunk 13]
```diff
@@ -0,0 +1,9 @@
+apiVersion: rbac.authorization.k8s.io/v1
+kind: ClusterRole
+metadata:
+  name: controller-role
```
Comment (baijum) [inline config/rbac/servicebinding_controller_role.yaml:4]: @bamachrn So, this should be changed to `service-binding-operator-controller-role`, right?

Comment (bamachrn) [reply config/rbac/servicebinding_controller_role.yaml:4]: yes

Comment (bamachrn) [reply config/rbac/servicebinding_controller_role.yaml:4]: my bad no, this gets prefixed with `service-binding-operator` by kustomaization while generating the bundle

File: config/rbac/servicebinding_controller_rolebinding.yaml (hunk 12 above)
Comment (pedjak) [reply config/rbac/servicebinding_controller_rolebinding.yaml:8]: but should not be the task of `kustomize` to prepend `service-binding-operator` prefix to this name?

File: test/acceptance/features/addRBACForOperator.feature [hunk 14]
```diff
@@ -0,0 +1,6 @@
+Feature: Making Service Biniding Operator RBAC pluggable so that other controllers/admins can add additional rules for the service binding operator.
```
Comment (pedjak) [inline test/acceptance/features/addRBACForOperator.feature:1]: can we name `.feature` file to reflect this - e.g. `pluggableRBAC.feature`?

File: test/acceptance/features/steps/openshift.py [hunk 15]
```diff
@@ -300,6 +305,18 @@ def get_resource_info_by_jsonpath(self, resource_type, name, namespace, json_pat
             print(f'Error getting value for {resource_type}/{name} in {namespace} path={json_path}: {output}')
             return None
 
+    def get_json_resource(self, resource_type, name, namespace):
+        error_msg = f"Error in getting resource: '{resource_type}' '{name}' namespace: '{namespace}'"
+        output, exit_code = self.cmd.run(f'{ctx.cli} get {resource_type} {name} -n {namespace} -o json')
+        assert exit_code == 0, error_msg
+        json_output = None
```
Comment (pedjak) [inline test/acceptance/features/steps/openshift.py:312]: you can remove this line.

File: test/acceptance/features/steps/openshift.py [hunk 16]
```diff
@@ -300,6 +305,18 @@ def get_resource_info_by_jsonpath(self, resource_type, name, namespace, json_pat
...
+        try:
+            json_output = json.loads(output)
+        except JSONDecodeError:
```
Comment (pedjak) [inline test/acceptance/features/steps/openshift.py:315]: we do not need to catch `JSONDecodeError` to throw another exception here. It is enough to do:

```python
json_output = json.loads(output)
assert json_output is not None, "Error parsing as JSON"
return json_output
```

File: config/rbac/servicebinding_controller_rolebinding.yaml (hunk 12 above)
Comment (bamachrn) [reply config/rbac/servicebinding_controller_rolebinding.yaml:8]: kustomization only changes `.metadata.name` not in the roleRef

File: test/acceptance/features/addRBACForOperator.feature (hunk 14 above)
Comment (bamachrn) [reply test/acceptance/features/addRBACForOperator.feature:1]: yeah makes sense

**Review by pedjak — APPROVED** [review]
